builtin/providers/azurerm/resource_arm_virtual_network_gateway.go

```diff
@@ -0,0 +1,685 @@
+package azurerm
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net/http"
+
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/arm/network"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceArmVirtualNetworkGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmVirtualNetworkGatewayCreate,
+		Read:   resourceArmVirtualNetworkGatewayRead,
+		Update: resourceArmVirtualNetworkGatewayCreate,
+		Delete: resourceArmVirtualNetworkGatewayDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"location": locationSchema(),
+
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.VirtualNetworkGatewayTypeExpressRoute),
+					string(network.VirtualNetworkGatewayTypeVpn),
+				}, false),
+				ForceNew: true,
+			},
+
+			"vpn_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(network.RouteBased),
+					string(network.PolicyBased),
+				}, false),
+				ForceNew: true,
+			},
+
+			"enable_bgp": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+
+			"active_active": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+
+			"sku": {
+				Type:     schema.TypeSet,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(network.VirtualNetworkGatewaySkuNameBasic),
+								string(network.VirtualNetworkGatewaySkuNameStandard),
+								string(network.VirtualNetworkGatewaySkuNameHighPerformance),
+								string(network.VirtualNetworkGatewaySkuNameUltraPerformance),
+							}, false),
+						},
+						"tier": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(network.VirtualNetworkGatewaySkuTierBasic),
+								string(network.VirtualNetworkGatewaySkuTierStandard),
+								string(network.VirtualNetworkGatewaySkuTierHighPerformance),
+								string(network.VirtualNetworkGatewaySkuTierUltraPerformance),
+							}, false),
+						},
+						"capacity": {
+							Type:     schema.TypeInt,
+							Computed: true,
+						},
+					},
+				},
+				Set: hashVirtualNetworkGatewaySku,
+			},
+
+			"ip_configuration": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
```

Comment:
I do not see how we can remove the required for the ip_configuration block. At least one is required to associate a public ip address (refer to https://docs.microsoft.com/en-us/cli/azure/network/vnet-gateway#create).

However, maybe I misunderstood your point.